File: StackWithLinkedList/stack.cpp
class StackNode
{
public:
  int data;
  StackNode *next;
};

StackNode *newNode(int data)
{
  StackNode *stackNode = new StackNode();
  stackNode -> data = data;
  stackNode -> next = NULL;
  return stackNode;
}

void push(StackNode **root, int data)
{
  StackNode *stackNode = newNode(data);
  stackNode -> next = *root;
  *root = stackNode;
}

int pop(StackNode **root)
{
  if (isEmpty(*root))
    return INT_MIN;

  StackNode *temp = *root;
  *root = (*root)->next;
  int popped = temp->data;
  delete temp;
  return popped;
}

int peek(StackNode *root)
{
  if (isEmpty(root))
    return INT_MIN;
  return root->data;
}

int isEmpty(StackNode *root) { return !root; }

int main()
{
  StackNode *root = NULL;
  push(&root, 1);
  push(&root, 2);
  push(&root, 3);
  pop(&root);
  peek(root);
}

File: StackWithLinkedList/stack.py
class StackNode:
  def __init__(self, data):
    self.data = data
    self.next = None


class Stack:
  def __init__(self):
    self.root = None

  def isEmpty(self):
    return True if self.root is None else False

  def push(self, data):
    newNode = StackNode(data)
    newNode.next = self.root
    self.root = newNode

  def pop(self):
    if (self.isEmpty()):
      return -1
    temp = self.root
    self.root = self.root.next
    popped = temp.data
    return popped

  def peek(self):
    if self.isEmpty():
      return -1
    return self.root.data


if __name__ == '__main__':
  stack = Stack()
  stack.push(1)
  stack.push(2)
  stack.push(3)
  stack.pop()
  stack.peek()

